book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,324 @@
+  * *As Above:* verifies each field in the block. Defers any data dependencies as
+    long as possible, awaits those data dependencies, then performs data
+    dependent checks.
+  * Note: Since futures are executed concurrently, we can use the same function
+    to:
+    * perform context-free verification,
+    * perform verification with deferred data dependencies,
+    * await data dependencies, and
+    * check data dependencies.
+    To maximise concurrency, we should write verification functions in this
+    specific order, so the awaits are as late as possible.
+* **State Updates:**
+  * *As Above:* the `BlockVerifier` returns success to the `ChainVerifier`,
+    which sends verified `Block`s to the state service.
+  * The state service implements the chain order consensus rule, which makes sure
+    each block is added to state after its previous block.
+
+## Zcash Protocol Design
+[zcash-protocol]: #zcash-protocol
+
+When designing a change to the Zcash protocol, minimise the data dependencies
+between blocks.
+
+Try to create designs that:
+* Eliminate data dependencies,
+* Make the changes depend on a version field in the block header or transaction,
+* Make the changes depend on the current Network Upgrade,
+* Make the changes depend on a field in the current block, with an additional
+  consensus rule to check that field against previous blocks, or
+* Prefer dependencies on older blocks, rather than newer blocks
+  (older blocks are more likely to verify earlier).
+
+Older dependencies have a design tradeoff:
+* depending on multiple blocks is complex,
+* depending on the previous block makes parallel verification harder.
+
+When making decisions about this dependency tradeoff, consider:
+* how the data dependency could be deferred, and
+* the CPU cost of the verification - if it is trivial, then it does not matter if
+  the verification is parallelised.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**TODO:** this design is a bit complicated, but we think it's necessary to
+          achieve our goals.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+**TODO:** expand on notes below
+
+- Why is this design the best in the space of possible designs?
+  - It's a good enough design to build upon
+- What other designs have been considered and what is the rationale for not choosing them?
+  - Serial verification
+  - Effectively single-threaded
+- What is the impact of not doing this?
+  - Verification is slow, we can't batch or parallelise some parts of the
+    verification
+
+# Prior art
+[prior-art]: #prior-art
+
+**TODO:**
+  - zcashd
+    - serial block verification
+    - Zebra implements the same consensus rules, but a different design
+  - tower
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- [ ] Is this design good enough to use as a framework for future RFCs?
+- [ ] Does this design require any changes to the current implementation?
```

Comment:
I think we can resolve this one immediately. as far as I can tell the answer is no
```suggestion
- [x] ~~Does this design require any changes to the current implementation?~~ No
```

Reply:
The design requires us to await the arrival of blocks in the state, rather than delay-polling in the BlockVerifier:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-consensus/src/block.rs#L188